core/src/main/java/org/axonframework/eventhandling/MultiEventHandlerInvoker.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2010-2018. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.eventhandling;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Implementation of {@link EventHandlerInvoker} with capabilities to invoke several different invokers.
+ *
+ * @author Milan Savic
+ * @since 3.3
+ */
+public class MultiEventHandlerInvoker implements EventHandlerInvoker {
```

Comment:
I assume you had to add this guy to allow a single `EventHandlerInvoker` to contain several `AbstractSagaManagers`, correct?

Reply:
Several event handler invokers - could be sagas, could be (simple) event handlers. Whatever you want :).

Reply:
But it's currently only used by the `AbstractSagaManagers` right?

Reply:
If you take a look at `DefaultEventProcessorRegistry` lines 184-188 you'll see that we are creating `MultiEventHandlerInvoker` with all `EventHandlerInvoker`s with the same processor name. That invoker could be `AnnotatedSagaManager` or `SimpleEventHandlerInvoker`.

Reply:
Ok, thanks for that reference.
That function you reference to makes it seem that we'll always wrap our event handler invokers in a `MultiEventHandlerInvoker`, regardless of the number of invokers created for a given Processing Group.
What's the use case for a `SimpleEventHandlerInvoker` to be wrapped by a `MultiEventHandlerInvoker`? In understand why we need it for our Sagas, but for the regular event handling approach I don't think we require that.

Reply:
Let's consider following case:
```
@ProcessingGroup("processor1") @Saga class Saga1 { }
@ProcessingGroup("processor1") @Saga class Saga2 { }
@ProcessingGroup("processor1") class EventHandler1 { }
@ProcessingGroup("processor1") class EventHandler2 { }
```
In the registry we would have one event processor with name `processor1` which would have one `eventHandlerInvoker` of type `MultiHandlerInvoker`. This invoker would have 3 delegates: 
 * `AnnotatedSagaManager` for `Saga1`
 * `AnnotatedSagaManager` for `Saga2`
 * `SimpleEventHandlerInvoker` with two event listeners: `EventHandler1` and `EventHandler2`.

Does this make sense to you or you think we should adapt the implementation somehow? (Since I'm not sure that I fully understand the underlying issue).

Reply:
Yes, it will be wrapped into `MultiEventHandlerInvoker`, I guess that we could do some kind of optimization here: if there is only one `SimpleHandlerInvoker` or `AnnotatedSagaManager` assigned to given processor name, we could avoid wrapping into `MultiEventHandlerInvoker`. What do you think?

Reply:
Yep that could be nice I think, makes for a slightly more intuitive solution when you'd for example be debugging through your Event Processors in a real live app.

But, what's your take on it. Would you find that optimization redundant @m1l4n54v1c?

Reply:
Then let's leave this as is and see what users come up with later on.